### PR TITLE
Block Bindings: Adapt block bindings to wordpress code latest changes

### PR DESCRIPTION
**Pull request by SantosGuillamot**
## What?
This pull request brings the latest changes included in WordPress core related to block bindings:
* [Add Block Bindings API helpers](https://github.com/WordPress/wordpress-develop/pull/5966).
* [Add block bindings registry class](https://github.com/WordPress/wordpress-develop/pull/5965).

## Why?
We need the block bindings API to follow the same format and make it compatible in the Gutenberg plugin.

## How?
I basically copied and paste the code from those PRs and make minor modifications to adapt it for Gutenberg.

## Testing Instructions

Ensure that the existing bindings keep working as expected:

0. Register a new custom field however you prefer. You can use a snippet similar to this:

```PHP
register_meta(
	'post',
	'text_custom_field',
	array(
		'show_in_rest' => true,
		'single'       => true,
		'type'         => 'string',
		'default'	   => 'This is the content of the text custom field',
	)
);
register_meta(
	'post',
	'url_custom_field',
	array(
		'show_in_rest' => true,
		'single'       => true,
		'type'         => 'string',
		'default'	   => 'https://wpmovies.dev/wp-content/uploads/2023/04/goncharov-poster-original-1-682x1024.jpeg',
	)
);
```

### In a page

**_Test paragraph_**
1. Add a paragraph with the content connected to a custom field:

```
<!-- wp:paragraph {"metadata":{"bindings":{"content":{"source":"core/post-meta","args":{"key":"text_custom_field"}}}}} -->
<p>Hello</p>
<!-- /wp:paragraph -->
```

2. Add a paragraph without any bindings.
3. Check that the paragraph with the bindings became non-editable.
4. Check that the paragraph shows the content of the custom field.
5. Check that the normal paragraph works as expected.
6. Go to the frontend and check that the value of the custom field is shown there.

**_Test heading_**

Repeat the paragraph test but using a heading.

**_Test button_**

1. Add a button with the text connected to a custom field.
2. Add another button with the URL connected to a custom field.
3. Add a button with the text and the URL connected to custom fields.

```
<!-- wp:buttons -->
<div class="wp-block-buttons"><!-- wp:button {"metadata":{"bindings":{"text":{"source":"core/post-meta","args":{"key":"text_custom_field"}}}}} -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button">TEXT</a></div>
<!-- /wp:button -->

<!-- wp:button {"metadata":{"bindings":{"url":{"source":"core/post-meta","args":{"key":"url_custom_field"}}}}} -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button" href="https://wpmovies.dev/wp-content/uploads/2023/03/3bhkrj58Vtu7enYsRolD1fZdja1-683x1024.jpg">URL</a></div>
<!-- /wp:button -->

<!-- wp:button {"metadata":{"bindings":{"text":{"source":"core/post-meta","args":{"key":"text_custom_field"}},"url":{"source":"core/post-meta","args":{"key":"url_custom_field"}}}}} -->
<div class="wp-block-button"><a class="wp-block-button__link wp-element-button" href="https://wpmovies.dev/wp-content/uploads/2023/03/3bhkrj58Vtu7enYsRolD1fZdja1-683x1024.jpg">TEXT</a></div>
<!-- /wp:button --></div>
<!-- /wp:buttons -->
```

4. Check that for the buttons with the text connected, they are not editable and they show the content of the custom field.
5. Check that for the buttons with the URL connected, the buttons to change the URL disappear from the UI.
6. Check that everything works in the frontend.

**_Test image_**

1. Add an image with the URL connected to a custom field.
2. Add an image with the alt attribute connected to a custom field.
3. Add an image with the title attribute connected to a custom field.
4. Add an image with more than one connection.

```
<!-- wp:heading {"level":3} -->
<h3 class="wp-block-heading">Image url and alt connected</h3>
<!-- /wp:heading -->

<!-- wp:image {"id":134,"sizeSlug":"large","linkDestination":"none","metadata":{"bindings":{"url":{"source":"core/post-meta","args":{"key":"page_url_custom_field"}},"alt":{"source":"core/post-meta","args":{"key":"page_text_custom_field"}}}}} -->
<figure class="wp-block-image size-large"><img src="https://wpmovies.dev/wp-content/uploads/2023/03/3bhkrj58Vtu7enYsRolD1fZdja1-683x1024.jpg" alt="Content of the page_text_custom_field" class="wp-image-134" title="Content of the PAGE text custom field"/></figure>
<!-- /wp:image -->

<!-- wp:heading {"level":3} -->
<h3 class="wp-block-heading">Alt connected</h3>
<!-- /wp:heading -->

<!-- wp:image {"sizeSlug":"large","linkDestination":"none","metadata":{"bindings":{"alt":{"source":"core/post-meta","args":{"key":"page_text_custom_field"}}}}} -->
<figure class="wp-block-image size-large"><img src="https://wpmovies.dev/wp-content/uploads/2023/04/q6y0Go1tsGEsmtFryDOJo3dEmqu-683x1024.jpg" alt="Content of the page_text_custom_field" title=""/></figure>
<!-- /wp:image -->

<!-- wp:heading {"level":3} -->
<h3 class="wp-block-heading">Nothing connected</h3>
<!-- /wp:heading -->

<!-- wp:image {"sizeSlug":"large","linkDestination":"none"} -->
<figure class="wp-block-image size-large"><img src="https://wpmovies.dev/wp-content/uploads/2023/04/q6y0Go1tsGEsmtFryDOJo3dEmqu-683x1024.jpg" alt=""/></figure>
<!-- /wp:image -->
```

5. For the images with the URL connected, check that the related buttons to link an image have disappeared and the image shows the URL from the custom field.
6. For the images with the alt/title connected, check that, in the right sidebar, they are disabled. Showing the value of the custom field instead and a message saying it is connected to custom fields.

### In a template

Go to a page template, for example, and repeat the process. In this case, the blocks can't show the value of the custom fields because it depends on each page. They should show a placeholder instead.

### Test pattern syncing overrides

1. Go to the Site Editor -> Patterns -> Create new synced pattern.
2. Add a paragraph, go to the Advanced section, and enable "Allow instance overrides".
3. Add a heading, go to the Advanced section, and enable "Allow instance overrides".
4. Add an image block, upload or select any media, go to the Advanced section, and enable "Allow instance overrides".
5. Add a button block, go to the Advanced section, and enable "Allow instance overrides".
6. Go to a page an insert the new pattern.
7. Modify the values of the blocks.
8. Check that the updated values are reflected in the frontend

### CODE REVIEW

File: lib/compat/wordpress-6.5/block-bindings/sources/pattern.php [hunk 1]
```diff
@@ -4,8 +4,8 @@
  *
  * @package gutenberg
  */
-if ( function_exists( 'wp_block_bindings_register_source' ) ) {
-	$pattern_source_callback = function ( $source_attrs, $block_instance, $attribute_name ) {
+if ( ! function_exists( 'gutenberg_register_block_bindings_pattern_overrides_source' ) && ! function_exists( 'gutenberg_block_bindings_pattern_overrides_callback' ) ) {
```
Comment (gziolo) [inline lib/compat/wordpress-6.5/block-bindings/sources/pattern.php]: You can remove these function checks. They are only necessary for functions that we expect to be defined outside the Gutenberg plugin. In our case, it's only for functions and classes defined in WordPress Core.

The same applies to the other file with the source.

File: lib/compat/wordpress-6.5/block-bindings/class-wp-block-bindings-registry.php [hunk 2]
```diff
@@ -0,0 +1,197 @@
...
+ * Block Bindings API: WP_Block_Bindings_Registry class.
+ *
+ * Support for overriding content in blocks by connecting them to different sources.
+ *
+ * @package WordPress
+ * @subpackage Block Bindings
+ * @since 6.5.0
+ */
+
+if ( class_exists( 'WP_Block_Bindings_Registry' ) ) {
```
Comment (gziolo) [inline lib/compat/wordpress-6.5/block-bindings/class-wp-block-bindings-registry.php:12]: Maybe try moving this check to the place where the class gets included - `load.php`.

```php
if ( class_exists( 'WP_Block_Bindings_Registry' ) ) {
    require __DIR__ . '/compat/wordpress-6.5/block-bindings/class-wp-block-bindings-registry.php';
}
```

Comment (SantosGuillamot) [reply lib/compat/wordpress-6.5/block-bindings/class-wp-block-bindings-registry.php:12]: Okay 🙂 Done in [this commit](https://github.com/WordPress/gutenberg/pull/58383/commits/056f09b72b9406d9646d63021ecb55f4ad8e0a12). Let's see if that solves the issue with unit tests.

File: lib/load.php [hunk 3]
```diff
@@ -112,10 +112,12 @@ function gutenberg_is_experiment_enabled( $name ) {
 require __DIR__ . '/compat/wordpress-6.5/interactivity-api/interactivity-api.php';
 require __DIR__ . '/compat/wordpress-6.5/class-wp-script-modules.php';
 require __DIR__ . '/compat/wordpress-6.5/scripts-modules.php';
-require __DIR__ . '/compat/wordpress-6.5/block-bindings/class-wp-block-bindings-registry.php';
 require __DIR__ . '/compat/wordpress-6.5/block-bindings/block-bindings.php';
 require __DIR__ . '/compat/wordpress-6.5/block-bindings/sources/post-meta.php';
 require __DIR__ . '/compat/wordpress-6.5/block-bindings/sources/pattern.php';
+if ( ! class_exists( 'WP_Block_Bindings_Registry' ) ) {
```
Comment (gziolo) [inline lib/load.php:115]: The class should be included before it is used. Like before this commit.

Comment (SantosGuillamot) [reply lib/load.php:115]: Done in [this commit](https://github.com/WordPress/gutenberg/pull/58383/commits/26aebb5fc0c118ff3d81a55e793049bb676819d5).

**Review by gziolo — APPROVED** [review]
Code changes look good! I haven’t tested, but it would be great to do some additional rounds of verification with trunk after landing.

File: lib/compat/wordpress-6.5/block-bindings/class-wp-block-bindings-registry.php [hunk 4]
```diff
@@ -0,0 +1,197 @@
+<?php
+/**
+ * Block Bindings API: WP_Block_Bindings_Registry class.
+ *
+ * Support for overriding content in blocks by connecting them to different sources.
+ *
+ * @package WordPress
```
Comment (michalczaplinski) [inline lib/compat/wordpress-6.5/block-bindings/class-wp-block-bindings-registry.php:7]:
```suggestion
 * @package Gutenberg
```

**Review by michalczaplinski — APPROVED** [review]
Looks good to me code-wise!

I've tested it for all the blocks and found no issues. 

I also tested using `trunk` of `wordpress-develop`. No issues either 👍 .

**Comment by cbravobernal** [comment]
I just cherry-picked this PR to the release/17.6 branch to get it included in the next release: 6042c9359f

**Pull request merged**